src/components/app/data/utils.js

```diff
@@ -747,3 +747,54 @@ export function isEnrollmentUpgradeable(enrollment) {
   const canUpgradeToVerifiedEnrollment = enrollment.mode === COURSE_MODES_MAP.AUDIT && !isEnrollByLapsed;
   return canUpgradeToVerifiedEnrollment;
 }
+
+export function determineAllocatedCourseRuns({
+  redeemableLearnerCreditPolicies,
+  courseKey,
+}) {
+  const { learnerContentAssignments } = redeemableLearnerCreditPolicies;
+  if (learnerContentAssignments.hasAllocatedAssignments) {
+    let allocatedCourseRunAssignments = learnerContentAssignments.allocatedAssignments.filter(
+      (assignment) => assignment?.isAssignedCourseRun,
+    );
+    if (courseKey) {
+      allocatedCourseRunAssignments = allocatedCourseRunAssignments.filter(
+        (assignment) => assignment?.parentContentKey === courseKey,
+      );
+    }
+    const allocatedCourseRunAssignmentKeys = allocatedCourseRunAssignments.map(assignment => assignment.contentKey);
+    const hasAssignedCourseRuns = allocatedCourseRunAssignmentKeys.length > 0;
+    const hasMultipleAssignedCourseRuns = allocatedCourseRunAssignmentKeys.length > 1;
+    return {
+      allocatedCourseRunAssignmentKeys,
+      allocatedCourseRunAssignments,
+      hasAssignedCourseRuns,
+      hasMultipleAssignedCourseRuns,
+    };
+  }
+  return {
+    allocatedCourseRunAssignmentKeys: [],
+    allocatedCourseRunAssignments: [],
+    hasAssignedCourseRuns: false,
+    hasMultipleAssignedCourseRuns: false,
+  };
+}
+
+export function transformCourseMetadataByAllocationCourseRun({
+  hasMultipleAssignedCourseRuns,
+  courseMetadata,
+  allocatedCourseRunAssignmentKeys,
+}) {
+  if (hasMultipleAssignedCourseRuns && allocatedCourseRunAssignmentKeys.length > 1) {
```

Comment:
[question/clarification] Why might we only want to mutate the below `courseRuns` and `availableCourseRuns` values when `hasMultipleAssignedCourseRuns: true`? Is there a particular reason we wouldn't want to go that if the user has a single allocated assignment applicable to the current course?

Reply:
When a user has a single allocated assignment, we alter how we retrieve course metadata.
``` jsx
  let courseRunKey = searchParams.get('course_run_key')?.replaceAll(' ', '+');
  if (!courseRunKey && hasAssignedCourseRuns) {
    courseRunKey = hasMultipleAssignedCourseRuns ? null : allocatedCourseRunAssignmentKeys[0];
  }
```

For a single allocation with no `course_run_key` defined, we are defaulting the `courseRunKey` to the allocated course run key, only returning a single course from course metadata. That way we don't have to go through the additional step of filtering by the allocated course run a second time.

For `hasMultipleAssignedCourseRuns: true` state, we are opting to use the `courseKey` to return all course runs from course metadata. Then we transform it further with this function. This approach allows us to treat the multiple course runs for a single course as a true edge case (that could easily be ripped out) as opposed to baking it into the default logic.